Makefile.toml

```diff
@@ -261,21 +261,33 @@ cargo install \
 '''
 ]
 
-[tasks.build-packages]
+# Builds a package including its build-time and runtime dependency packages.
+[tasks.build-package]
 dependencies = ["build-tools", "publish-setup"]
+script_runner = "bash"
 script = [
 '''
+set -e
+if [ -z "${PACKAGE}" ]; then
+    echo "the PACKAGE environment variable must be set. for example:"
+    echo "cargo make build-package -e PACKAGE=kernel"
+    exit 1
+fi
```

Comment:
Consider this scenario. Note that I haven't tested this, so I may be wrong in some of the specifics, but I expect I'm right about the larger issue.
```
# Get to a fresh state
cargo make clean

# builds the kernel and os package for the 1.18 variant crate
cargo make -e BUILDSYS_VARIANT=aws-k8s-1.18

# rebuilds the kernel and os packages for the 1.19 variant crate,
# since cargo doesn't know the RPMs exist yet, even though the
# kernel package is not variant-specific.
cargo make -e BUILDSYS_VARIANT=aws-k8s-1.19

# builds the kernel package for the packages workspace,
# resulting in a new RPM build that doesn't match what we
# used for either 1.18 or 1.19.
cargo make -e BUILDSYS_VARIANT=aws-k8s-1.18 repo

# reuses the kernel package build from the packages workspace.
cargo make -e BUILDSYS_VARIANT=aws-k8s-1.19 repo

# reuses the kernel and os package builds for the 1.19 crate,
# which is incorrect since "kernel" has been rebuilt, possibly
# with a different config, and more critically because "os" has
# been rebuilt. cargo won't trigger a rebuild since none of
# the source files have changed and the environment matches
# what it was for the last 1.18 build.
cargo make -e BUILDSYS_VARIANT=aws-k8s-1.18

# reuses the kernel and os package builds for the 1.19 crate,
# which is incorrect since "kernel" has been rebuilt, possibly
# with a different config.
cargo make -e BUILDSYS_VARIANT=aws-k8s-1.19
```

The underlying problem is that we have only a single cache directory for RPMs, but `cargo` has separate views of that cache depending on which workspace or crate was last built.

I am not sure of the best way to fix it. The `build-package` problem may not be solvable if we take the position that the kernel package that we use to generate the kmod kit - or more generally, any package that contributes a repo artifact - must exactly match the package we used to build the image. Essentially this means the separate packages workspace might have to go away, leaving us with just the variant crates.

For detecting that "os" RPMs have been silently rebuilt, it might work to emit `rerun-if-changed` directives on the RPM and image outputs. I am not sure it's sufficient but it's definitely an improvement.

Otherwise we might have to unify all variants back into a single workspace, so that there's no mismatch between cargo's view of the most recent builds and the contents of our `build` directory.

Reply:
Ugh, I don't quite follow but I'll look more closely at this. It's hard to wrap my head around which system governs rebuilding since there are four (?) competing systems of build caching.
- cargo
- rpm
- docker buildkit
- buildsys

In my mind everything would 'just work' because cargo doesn't know about rpms and only rebuilds based on a variant change if we tell it to. Cargo has the ultimate veto power in terms of re-building something, no? I mean if cargo doesn't need to rebuild the empty lib.rs (actually pkg.rs) then it won't rebuild it right?

Reply:
P.S. rpms that change how/what they build based on variant wasn't something I was considering.

Reply:
I think I understand the problem now.

The flaw in my thinking was that I didn't account for the fact that we used to have a shared `target` directory for all packages because we were using the `packages` workspace. In my current solution we have different `target` directories for each variant, and a different one still when we use the packages workspace to `build-kernel` for the kmod kit.

Reply:
This is now fixed in 76cb1ad by pointing a cargo build to the workspace Cargo.toml